### ISSUE TITLE
EMS parser problem with negative number on right side of IF

### ISSUE THREAD

**Issue by mjwitte** [NREL/EnergyPlus#5257]
Helpdesk ticket 10812
From user
Implementing an EMS model I encoutered the following issue.

ELSEIF TPrim == -1, 
Gives the following output in *.edd when TPrim = 16.
AIRPRIMZ01,Line 15,ELSEIF TPRIM == -1,16.000000, Occurrence info=DESIGNCOOL, 07/15 12:15 - 12:30
So that the condition is actually assessed true (>0). 

ELSEIF -1 == TPrim , 
Gives the correct output (0).
AIRPRIMZ02,Line 15,ELSEIF -1 == TPRIM,0.0, Occurrence info=DESIGNCOOL, 07/15 19:45 - 20:00

Workaround:
If you make the expression: (0 - 1)  Instead of  -1  It should work.

No defect file was attached to the helpdesk ticket.  Will need to make one.

**Comment by rraustad** [comment]
Unit test reflects above description.
Substituting each of the commented lines yields:
1st IF passes ( IF MyVar == 1 )
2nd IF fails ( IF MyVar == -1 )
3rd IF passes ( IF -1 == MyVar )
4th IF passes ( IF MyVar == 0-1 )

```
TEST_F( EnergyPlusFixture, Test_EMSLogic ) {

std::string const idf_objects = delimited_string( {
    "Version,8.4;",

    "OutdoorAir:Node, Test node;",

    "EnergyManagementSystem:Actuator,",
    "TempSetpointLo,          !- Name",
    "Test node,  !- Actuated Component Unique Name",
    "System Node Setpoint,    !- Actuated Component Type",
    "Temperature Minimum Setpoint;    !- Actuated Component Control Type",

    "EnergyManagementSystem:Actuator,",
    "TempSetpointHi,          !- Name",
    "Test node,  !- Actuated Component Unique Name",
    "System Node Setpoint,    !- Actuated Component Type",
    "Temperature Maximum Setpoint;    !- Actuated Component Control Type",


    "EnergyManagementSystem:ProgramCallingManager,",
    "Dual Setpoint Test Manager,  !- Name",
    "BeginNewEnvironment,  !- EnergyPlus Model Calling Point",
    "DualSetpiontTestControl;  !- Program Name 1",

    "EnergyManagementSystem:Program,",
    "DualSetpiontTestControl,",
    "Set MyVar = 10,",
//      "IF MyVar == 1,", // pass
//      "IF MyVar == -1,", // fail
//      "IF -1 == MyVar,", // pass
    "IF MyVar == 0-1,", // pass
    "Set TempSetpointLo = 11.0,",
    "Set TempSetpointHi  = 12.0,",
    "ELSE,",
    "Set TempSetpointLo = 16.0,",
    "Set TempSetpointHi  = 20.0,",
    "ENDIF;",

} );

ASSERT_FALSE( process_idf( idf_objects ) );

OutAirNodeManager::SetOutAirNodes();
EMSManager::CheckIfAnyEMS();
EMSManager::FinishProcessingUserInput = true;
EMSManager::ManageEMS( DataGlobals::emsCallFromSetupSimulation );
EMSManager::ManageEMS( DataGlobals::emsCallFromBeginNewEvironment );

EXPECT_NEAR( DataLoopNode::Node( 1 ).TempSetPointHi, 20.0, 0.000001 );
EXPECT_NEAR( DataLoopNode::Node( 1 ).TempSetPointLo, 16.0, 0.000001 );

}
```

**Comment by rraustad** [comment]
@Myoldmopar @EnergyArchmage @mjwitte 

I have a handle on this now.

There are several possibilities for a negative number and many I probably haven't considered.
There are also at least 2 configurations of EMS commands that affect this issue, the SET and IF commands.

The simplest of these are handled [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L1140).

Start with these examples:
SET Myvar = -1;
IF -1 == Myvar;

The code will strip the IF command through the first space (' ') [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L497) and the SET command through the '=' [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L520) leaving the remaining expression, for each example, as:

-1
-1 == MyVar

Then in ParseExpression the code [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L1140) converts the negative sign to a subtract sign and then parses as a normal expression as:

0-1
0-1 == MyVar

These strings get stored as they are parsed into the Token array (first example 1-3, and second example 1-5).

Token(1).String = 0, Token(1).Type = TypeNumber
Token(2).String = -, Token(2).Type = TypeOperator, Token(2).Operator = OperatorSubtract
Token(3).String = 1, Token(3).Type = TypeNumber
Token(4).String = =, Token(4).Type = TypeOperator
Token(5).String = MyVar, Token(5).Type = TypeVariable

I hope I got that right from memory, but you get this idea. Each element or token of the "remainder" of the string is passed as an expression and parsed and stored in the Token array.

Now the issue with an example.

IF MyVar == -1;

The "expression" for the IF is (stripped to the first space):
MyVar == -1
so the code to fix the negative sign breaks down since the negative sign is within the string instead of at the beginning. There was an attempt to fix this using the logicals MinusFound and OperatorProcessing [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L1167-L1169), [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L1237) and [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L1601), except that it didn't execute correctly because OperatorProcessing was reset [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L1169) instead of outside this while loop and then handled correctly within the while loop via manual reset.

Since the negative sign was never process correctly, the number (e.g., -1) was not [inverted](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L1237) and instead the negative sign was processed as a Token and the 1 processed as a second Token (there should have only been 1 Token for this as the numeral -1 after inverting the sign) and the IF test fails (always false I think).

So now to the fix. I think the best fix for this "might" be to insert a 0 in the string where appropriate and parse the string normally. I did attempt to process the negative sign but I think that would fail if the negative sign where not directly after an operator, i.e., the ==, ( or other, so I could set the OperatorProcessing logical.

This should already work when I fix the MinusFound and OperatorProcessing issue:
IF MyVar == -@PI

```
UPDATE: thinking more of this, I don't think a minus sign will work here since this won't apply because @PI is not a numberal:
if ( MinusFound ) Token( NumTokens ).Number = -Token( NumTokens ).Number;
    A 0 would have to be inserted to make this 0-@PI
```

But this will fail:
IF MyVar == 10 \* -@PI

Since MyVar == 10 \* 0-@PI should get evaluated as (10 \* 0)-@PI

But this would work?
IF MyVar == 10 \* (-@PI)

And get converted to:
MyVar == 10 \* (0-@PI)

If we insert a 0 I think the parser would get this correct.

So you can see there are many possible formulations of a negative sign. I can only hope to get the majority of simple math expressions at this point.

I just had trouble last night trying to figure out how to stuff a "0" in the middle of a string.

The string is pass as a pointer [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.hh#L214) and I wasn't sure how to handle modifying a string that was passed as a pointer.

So if someone could help me modify the string, I could set up the changes on a branch and then someone could follow up with correcting my sloppy C++ programming. At least the follow up would not have to go through what I have the last few nights. They could instead start with a map of where the code needed revision.

For example, right [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L1219) I'd like to insert a 0 into the middle of the String. Then manually process the 0 as a Token (i.e., set up StringToken = "0", etc.), allowing the parser to pick up from there.

The compiler would not let me do this:
String = String[0, Pos] + "0" + String[Pos+1, LastPos + 1]

Interesting, the compiler didn't complain when I just tried this but it did crash at this line in the debugger. So what is the proper syntax?  Maybe I need a local string to manipulate. Hopefully that would work and the original String would not be used elsewhere (I don't know that yet).

I'd like to get this working on a branch at least to a point where someone else can come in behind and clean it up.

I think I got this to work last night, but then it failed again in a more complex unit test so there is something else that needs to be addressed.

```
"Set MyVar = -1,",
"IF MyVar == -1,", // pass
```

Pardon the multiple links, I wanted to document this as best I could if I wind up not able to complete this.

I will also add that there is another possible solution since this was already thought of but not finished. This is commented out in function SetupPossibleOperators.

```
    // not sure how to distinguish from subtract in parsing of tokens, not yet available
    //  PossibleOperators(OperatorNegative)%NumOperands = 1
    //  PossibleOperators(OperatorNegative)%Code        = OperatorNegative
    //  PossibleOperators(OperatorNegative)%Symbol      = '-'
```

So an alternative would be to get this to work and not need to hack into the code by modifying the string. The issue of knowing when the - is a negative sign or a subtract sign may still be an issue. I haven't thought that far ahead.

**Comment by rraustad** [comment]
Even if I correct the math in my previous post this doesn't work.

Remember that Pos holds the - and I want to insert before the -

String = String[ 0, Pos -1 ] + "0" + String[ Pos, LastPos ];

What is needed to change this string? Do I need to redimension the String to include an extra character?

**Comment by rraustad** [comment]
One more documenting item. The Token counter is always incremented each time through the while loop [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L1162). So for code like [this](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L1225-L1226), either the Token counter NumTokens needs to be decremented, the Token array needs to be redimensioned, and the last line [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/RuntimeLanguageProcessor.cc#L1655) cannot be executed, or else something needs to get processed into the Token array.

**Comment by Myoldmopar** [comment]
Closed via #5348
